NWNXLib/Services/PerObjectStorage/PerObjectStorage.cpp

```diff
@@ -0,0 +1,224 @@
+#include "Services/PerObjectStorage/PerObjectStorage.hpp"
+#include "API/Constants.hpp"
+
+namespace NWNXLib {
+
+namespace Services {
+
+std::unordered_map<API::Types::ObjectID, std::unique_ptr<PerObjectStorage::ObjectStorage>> PerObjectStorage::g_objectStorage;
+
+PerObjectStorage::ObjectStorage& PerObjectStorage::GetObjectStorage(API::Types::ObjectID object)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+        return *it->second;
+
+    return *g_objectStorage.emplace(object, std::make_unique<PerObjectStorage::ObjectStorage>(object)).first->second;
+}
+
+
+void PerObjectStorage::Set(API::Types::ObjectID object, std::string key, int value)
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetIntMap().emplace(key, value);
+}
+void PerObjectStorage::Set(API::Types::ObjectID object, std::string key, float value)
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetFloatMap().emplace(key, value);
+}
+void PerObjectStorage::Set(API::Types::ObjectID object, std::string key, std::string value)
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetStringMap().emplace(key, value);
+}
+void PerObjectStorage::Set(API::Types::ObjectID object, std::string key, void *value, CleanupFunc cleanup)
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetPointerMap().emplace(key, std::make_pair<>(value, cleanup));
```

Comment:
You don't need the brackets on `make_pair<>` - it's fine to do `make_pair(first, second)` as the types are deduced automatically. Doesn't matter either way - just thought you might want to know!